tests/python/test_py_dataclass.py

```diff
@@ -864,3 +864,32 @@ def k1(p1: P1) -> None:
     assert d[0] == 55
     assert e[0] == 66
     assert f[0] == 77
+
+
+@test_utils.test()
+def test_template_mapper_cache():
+    gc.collect()
+    gc.collect()
+
+    @dataclass(frozen=True)
+    class MyStruct:
+        value: ti.types.ndarray()
+
+    @ti.kernel
+    def my_kernel(my_struct_1d: MyStruct, my_struct_2d: MyStruct) -> None:
+        for i in ti.ndrange(my_struct_1d.value.shape[0]):
+            my_struct_1d.value[i] += 1
+        for i, j in ti.ndrange(my_struct_2d.value.shape[0], my_struct_2d.value.shape[1]):
+            my_struct_2d.value[i, j] += 1
+
+    value = ti.ndarray(ti.i32, shape=(1,))
+    value.fill(0)
+    my_struct_1d = MyStruct(value=value)
+    value = ti.ndarray(ti.f32, shape=(1, 2))
+    value.fill(0.0)
+    my_struct_2d = MyStruct(value=value)
+
+    my_kernel(my_struct_1d, my_struct_2d)
+    assert my_struct_1d.value[0] == 1
+    assert abs(my_struct_2d.value[0, 0] - 1.0) < 1e-9
```

Comment:
Can we define EPS, rather than writing 1e-9 each time please?

Also, looks like python provides `maths.isclose`, https://docs.python.org/3/library/math.html#math.isclose Thoughts on our using that?

Reply:
No tolerance is needed indeed. I will remove it if you think it would be better.